staging/src/k8s.io/apiserver/pkg/storage/cacher/store_btree.go

```diff
@@ -0,0 +1,404 @@
+		items = append(items, i.(interface{}))
+		return true
+	})
+
+	return items
+}
+
+func (s *btreeStore) ListKeys() []string {
+	items := make([]string, 0, s.tree.Len())
+	s.tree.Ascend(func(i btree.Item) bool {
+		items = append(items, i.(*storeElement).Key)
+		return true
+	})
+
+	return items
+}
+
+func (s *btreeStore) Get(obj interface{}) (item interface{}, exists bool, err error) {
+	storeElem, ok := obj.(*storeElement)
+	if !ok {
+		return nil, false, fmt.Errorf("obj is not a storeElement")
+	}
+	item = s.tree.Get(storeElem)
+	if item == nil {
+		return nil, false, nil
+	}
+
+	return item, true, nil
+}
+
+func (s *btreeStore) GetByKey(key string) (item interface{}, exists bool, err error) {
+	return s.getByKey(key)
+}
+
+func (s *btreeStore) Replace(objs []interface{}, _ string) error {
+	s.tree.Clear(false)
+	for _, obj := range objs {
+		storeElem, ok := obj.(*storeElement)
+		if !ok {
+			return fmt.Errorf("obj not a storeElement: %#v", obj)
+		}
+		s.addOrUpdateElem(storeElem)
+	}
+	return nil
+}
+
+// addOrUpdateLocked assumes a lock is held and is used for Add
+// and Update operations.
+func (s *btreeStore) addOrUpdateElem(storeElem *storeElement) *storeElement {
+	oldObj := s.tree.ReplaceOrInsert(storeElem)
+	if oldObj == nil {
+		return nil
+	}
+	return oldObj.(*storeElement)
+}
+
+func (s *btreeStore) getByKey(key string) (item interface{}, exists bool, err error) {
+	keyElement := &storeElement{Key: key}
+	item = s.tree.Get(keyElement)
+
+	return item, item != nil, nil
+}
+
+func (s *btreeStore) ListPrefix(prefix, continueKey string, limit int) ([]interface{}, bool) {
+	if limit < 0 {
+		return nil, false
+	}
+	if continueKey == "" {
+		continueKey = prefix
+	}
+	var result []interface{}
+	var hasMore bool
+
+	if limit == 0 {
+		s.tree.AscendGreaterOrEqual(&storeElement{Key: continueKey}, func(i btree.Item) bool {
+			elementKey := i.(*storeElement).Key
+			if !strings.HasPrefix(elementKey, prefix) {
+				return false
+			}
+			result = append(result, i.(interface{}))
+			return true
+		})
+		return result, hasMore
+	}
+
+	s.tree.AscendGreaterOrEqual(&storeElement{Key: continueKey}, func(i btree.Item) bool {
+		elementKey := i.(*storeElement).Key
+		if !strings.HasPrefix(elementKey, prefix) {
+			return false
+		}
+		// TODO: Might be worth to lookup one more item to provide more accurate HasMore.
+		if len(result) < limit {
+			result = append(result, i.(interface{}))
+		} else {
+			hasMore = true
+			return false
+		}
+		return true
+	})
+
+	return result, hasMore
+}
+
+func (s *btreeStore) Count(prefix, continueKey string) (count int) {
+	if continueKey == "" {
+		continueKey = prefix
+	}
+	s.tree.AscendGreaterOrEqual(&storeElement{Key: continueKey}, func(i btree.Item) bool {
+		elementKey := i.(*storeElement).Key
+		if !strings.HasPrefix(elementKey, prefix) {
+			return false
+		}
+		count++
+		return true
+	})
+
+	return count
+}
+
+func newIndexer(indexers cache.Indexers) indexer {
```

Comment:
This is almost exactly a copy-paste of `storeIndex` from client-go/cache.

I would much rather prefer to make that public and simply use it here instead of having two almost identical implementation of the same thing.

Reply:
It's almost identical, but it's not. I index to object, not object key. The reason is that serving from index with keys requires a Get to store. For client-go/cache Get is cheap as it's a map. For btree individual lookup of hundreds of keys is expensive. 

To maintain same performance for indexed reads I changed the indexer to point to values and not key.

Reply:
OK - I see. That makes sense, but I don't think it changes the core of my comment.

It still seems relatively simple to refactor the existing indexer, by simply parametrizing it with the "ObjectCollection` (we need a better name).
- ObjectCollection implements simple interface of Put(key, value) & Delete(key, value)
- the default implementation uses sets.String (we just need a simple wrapper for it) and remains as is
- your implementation uses map[string]interface{} (or map[string]*storeElement)

Reply:
It does due to https://github.com/kubernetes/kubernetes/pull/126754#discussion_r1735760083
There is a different semantics of having indexes to
* to unique immutable key
* to mutable object value

The difference is in how update to object is implemented. In first case we don't care, key value is immutable, however in second case we need to update on each value change. That's why https://github.com/kubernetes/kubernetes/pull/126754#discussion_r1735760083 is important change.

Reply:
> The difference is in how update to object is implemented. In first case we don't care, key value is immutable, however in second case we need to update on each value change. That's why https://github.com/kubernetes/kubernetes/pull/126754#discussion_r1735760083 is important change.

This is easily fixable - you add `Update` method to the interface I mentioned and Update is no-op in a default implementation and is what you have in the btree-store implementation.

Reply:
Just worried that we might be doing code dedup, just for sake of deduping code.

Reply:
> Just worried that we might be doing code dedup, just for sake of deduping code.

historically almost every such structure had some subtle bugs (IIRC including that one).
Everyone likes writing new code - we should just start reusing what we have instead of having N different ways of doing almost the same thing.

Reply:
> historically almost every such structure had some subtle bugs (IIRC including that one).

Maybe the reason was lack of tests? Looking at them, they leave a lot to be desired. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/thread_safe_store_test.go. 

Don't follow your argument about liking to write new code, I took the original code by @MadhavJivrajani and really only coded the tests.

My point stands, unifying code just for deduplication sake can have devastating consequences, where you force unified behavior, by hacking around misalignments, just to have one solution. The real problem appears, when it turns out that the use cases you combined are totally different, driven by different users and different needs, causing overhead on both.

To give one argument why duplication can be good, please read https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

Reply:
OK - I don't want to block this change, so I can live with it if you add a TODO to consider unifying the code in the follow up.

Reply:
Can you please add such TODO at class defiinition?

Reply:
Done